### DISCUSSION TITLE
Unable to Run FRRouting with CORE EMULATOR (Using VIRTUAL BOX with OS UBUNTU 20.04.1 LTS)

### DISCUSSION THREAD

**Discussion by abdul152** [coreemu/core#536] [Q&A]
I just installed CORE emulator. FRR services is showing on every node inside the CORE emulator. But these are not working. I ran these services and try to use vtysh then configuration prompt is not appearing. I tried to click on nodes and the go to bash window to type vtysh so I get error that "no daemons found". However, Quagga is working properly.
Please help me out.
All this setup is running on Virtual box Vm with OS Ubuntu 18.04 LTS Desktop machine.
Thanks in advance.
![1](https://user-images.githubusercontent.com/56837476/101122165-1f64f580-35f2-11eb-9820-fc6e9e7333d0.jpg)
![2](https://user-images.githubusercontent.com/56837476/101122166-1ffd8c00-35f2-11eb-88d8-7855b4347d0a.jpg)

**Comment by bharnden** [comment]
You say Quagga is working perfectly? Does that mean you have both FRR and Quagga installed by chance? I do not think that is desirable.

I also see you mention both Ubuntu 20 and 18. If you are using a WLAN, Ubuntu 20 will have issues, unless you install the legacy ebtables support.

Beyond that, I do not know offhand if FRR updates have deviated from the CORE FRR services that were contributed, which may be a possibility.

**Comment by bharnden** [comment]
You could open a node that should run FRR after startup and attempt to run the generated script manually to see if any errors occur, which may help narrow down the specifics of your problem. Or look at the running process etc, to see if things appear as expected.

**Comment by abdul152** [comment]
For Quagga, I am sure it is running
But, for FRR I am not sure. But, in the services section of node I can see both quagga and FRR. 
If I configure with quagga it is working fine and I can take access via vtysh. But when I am trying with FRR then, vtysh not working.

**Comment by bharnden** [comment]
That didn't answer the question are you installing both FRR and Quagga? FRR is a fork of Quagga from my understanding, so you would NOT want both installed, it is one or the other.

**Comment by abdul152** [comment]
I have not expertise with it. But, my need is that only FRR can work with CORE.

**Comment by bharnden** [comment]
I forget that by default now OSPF MDR is installed, which is a variation of Quagga. You would likely need to uninstall this first. Then install FRR to be safe.

```shell
cd <CORE REPO PATH>
cd ../ospf-mdr
sudo make uninstall
```

Then you would install FRR however you installed it, if you did not, you could follow their directions:
http://docs.frrouting.org/en/latest/installation.html

**Comment by abdul152** [comment]
Thank you so much for immediate help. 
I'll ping you in the morning if this works.

**Comment by abdul152** [comment]
I have one more question. After uninstalling this when I will install FRR then it will automatically integrate with CORE Emulator ? Or else after installaltion of FRR I have to do something ?

**Comment by bharnden** [comment]
CORE only has support by way of the FRR services, which are just one potential configuration for running FRR routing functionality.

The service would generate files for start FRR services, which the generated configuration. FRR likely supports a multitude of configuration options, which is likely within their documentation. So if you wanted anything specific or different about the configuration to tailor it to your needs you would need to modify the service and the files it needs or set it all up yourself.

The FRR service and more or less all CORE services serve as a convenience for basic functionality. If you need custom configuration or behavior, that will always result in modifying a service, creating a custom service, or manually driving your desired efforts by hand or programmatically. Any and all of that is possible, assuming you know what you need to do.

**Comment by bharnden** [comment]
A better way to phrase it, is that services are not really deep integrations. They are just conveniences for running shell scripts for a purpose, and potentially can dynamically leverage information for a node running a service, to help make these scripts context aware.

**Comment by abdul152** [comment]
Thank you so much. 
Basically, I have cisco background and I have nlt much expertise with CORE and FRR. I need help to install FRR with CORE. After installation I have to run an MPLS environment based on OSPF/ISIS, LDP and MP-BGP for Layer 2 and layers 3 VPNS. 

If you can guide me regarding the installation of FRR with CORE so, I can proceed it ahead.

**Comment by bharnden** [comment]
CORE nodes allow you to run software installed on the system, with some trade offs for needing isolation for certain directories. Which is an are services can help alleviate.

But to install FRR, you would look to install it how you normally would, following the link I posted above. Any further delving into setting up FRR and the things you described would not really be my area to help. That is getting more into the specifics of the software you are running.

FRR documentation etc, would be your friend at that point.

**Comment by abdul152** [comment]
Hi Bharnden, 
Somehow, FRR now working with CORE and vtysh promt is also appearing. But now I am facing one more issue. For example I configure four nodes with FRRisis and FRRzebra. and then run the topology. At first its working fine. But If I configure something on any router it will be save on /etc/frr/frr.conf and /etc/frr/frr.conf. So, when I stop the services and restart all of these so every node got 2 IPs on every interface and I dont know why. 
Can you may be help me in this regard too.

**Comment by bharnden** [comment]
If you look into the example service for FRR, it appears to expect things using the `/usr/local` prefix and it tries to compensate by symlinking to the directories you mentioned. If you actions are overwriting these files or they were created by your installation, then that is likely complicating your assumptions.

https://github.com/coreemu/core/blob/master/daemon/core/services/frr.py

This is all getting into the nuance of the using FRR directly, how you are trying to use it, and the file system. 

Know that CORE nodes are all sharing the same filesystem, so if configuration files are not isolated they a given node by way of a service defining directories, they will use the same files resulting it odd behavior.

You can see the example for the FRR service and what it is isolating by the following line:
https://github.com/coreemu/core/blob/a57b838f191640a49e45347a1764bf05b3269e0a/daemon/core/services/frr.py#L22

I cannot really help much more into FRR specifics, as that is not within my knowledge to help, potentially their documentation will help you.
